Console command support (:use, :params, :clear, history)

OskarDamkjaer opened the pull request:
I introduce a new grammar that extends the `CypherParser` with more client side commands like `:param x => 23` and `:use foo`. It can be toggled on/off, and I've added proper errors, completions, highlighting for the client side commands. See the new test cases for a fuller picture

This PR also fixes semantic analysis for multi-statement queries, like `RETURN a; RETURN b;` since I had to split the string into parts to avoid sending the client commands regardless.

ncordon commented on packages/language-support/package.json:
```diff
@@ -47,7 +47,7 @@
   },
   "scripts": {
-    "gen-parser": "antlr4 -Dlanguage=TypeScript -visitor src/antlr-grammar/CypherParser.g4 src/antlr-grammar/CypherLexer.g4 -o src/generated-parser/ -Xexact-output-dir",
+    "gen-parser": "antlr4 -Dlanguage=TypeScript -visitor src/antlr-grammar/CommandLexer.g4 src/antlr-grammar/CommandParser.g4 -o src/generated-parser/ -Xexact-output-dir",
```
This is a nitpick but could we find another name for the files? `CypherCmdParser`?

ncordon reviewed (commented):
I also find a bit strange that in VSCode we give you an error saying you need to complete with a console command, not help you in the auto-completion but then error saying you cannot use these type of commands here.

I think it would make more sense to show the error, but still give you the auto-completion

<img width="499" src="https://github.com/neo4j/cypher-language-support/assets/5649971/6ecdbfef-1267-4f94-9a4f-d2853238a774">

<img width="561"  src="https://github.com/neo4j/cypher-language-support/assets/5649971/da475d43-d902-4bb9-8910-3ab8ab2296a2">

ncordon reviewed (commented):
Also another bug, but this blows up VSCode, not sure if related to the PR:

<img width="926"  src="https://github.com/neo4j/cypher-language-support/assets/5649971/0bdb70d1-c0d3-49cb-8da9-7e961194cc01">

OskarDamkjaer commented on packages/language-support/src/antlr-grammar/CommandParser.g4:
```diff
@@ -0,0 +1,28 @@
+
+historyCmd: HISTORY;
+
+useCmd: useCompletionRule symbolicAliasName?;
```
Yes! But also because the semantics are different. USE in the database allows `USE graph.byName('myComposite.myConstituent')` which we don't allow in `:use`

OskarDamkjaer replied:
We also allow running it without passing an argument, which switches you to the default database

OskarDamkjaer commented on packages/react-codemirror/src/lang-cypher/autocomplete.ts:
```diff
@@ -28,7 +28,9 @@ const completionKindToCodemirrorIcon = (c: CompletionItemKind) => {
     [CompletionItemKind.Struct]: 'Struct',
-    [CompletionItemKind.Event]: 'Event',
+    // we're missuing the enum here as there is no Console command in the item kind list
+    // deceide which one we want to use - for vscode we can't control the icon
```
I want to get rid of the second part of the comment altogether. let's stick with "event" for now and switch it if we dislike the icon

OskarDamkjaer commented on packages/language-support/src/highlighting/syntaxValidation/syntaxValidation.ts:
```diff
@@ -169,21 +175,30 @@ export function validateSyntax(
 
+    // semantic anaylsis doesn't handle multi statements
+    // we break the file into statements and run semantic analysis on each
+    // then mapp the postions back to the original file
```
I updated/rephrased the comment all together

OskarDamkjaer commented:
@ncordon the odd behavior you saw in vsocde/codemirror was caused by the `parseToCommands` crashing on parse errors. I fixed the errors and added a few tests.

Is it fine to leave in the issue of the error message suggesting a keyword on `:` for now and re-visit it if it comes up again?

ncordon reviewed (changes requested):
I think it's still not working correctly and there are a few details to polish.

When I am on VSCode, I get the error on `:PARAM`, but the moment I write something else, that error disappears and I get no auto-completions for normal cypher keywords:

<img width="567" src="https://github.com/neo4j/cypher-language-support/assets/5649971/f66be43f-82a9-4f10-bf9c-12884fb3efd3">

And:
<img width="315" src="https://github.com/neo4j/cypher-language-support/assets/5649971/60dca7e3-8c7a-47da-b02f-6b4ae4a817dd">
